### DISCUSSION TITLE
Что-то непонятное с очередью на БГ

### DISCUSSION THREAD

**Discussion by Bogir** [trickerer/Trinity-Bots#907] [General]
В очередь на БГ встало 16 ботов из 10 возможных. Свободных ботов около 500 +/-.
![изображение](https://github.com/user-attachments/assets/5a07c366-51b4-4a7a-8397-ae2d3ed62999)
![изображение](https://github.com/user-attachments/assets/a46f5bfd-3c1f-484b-a42a-7f758c5d74fe)

6 лишних отсеялись
![изображение](https://github.com/user-attachments/assets/c96b544c-26f1-490f-9394-d069f16a9bd6)

Почему эти 10 делятся на 2 команды по 5 ботов в каждую? 
![изображение](https://github.com/user-attachments/assets/7e77a6df-a0e2-43b5-baa5-b9f5c82ed683)

Может правильнее было бы набирать во вторую очередь тоже столько же ботов за противоположную фракцию? Или как там это устроено?

**Comment by trickerer** [comment]
> В очередь на БГ встало 16 ботов из 10 возможных. Свободных ботов около 500 +/-. ![изображение](https://private-user-images.githubusercontent.com/25766571/378899931-5a07c366-51b4-4a7a-8397-ae2d3ed62999.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MzgyMTQsIm5iZiI6MTcyOTYzNzkxNCwicGF0aCI6Ii8yNTc2NjU3MS8zNzg4OTk5MzEtNWEwN2MzNjYtNTFiNC00YTdhLTgzOTctYWUyZDNlZDYyOTk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDIyNTgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2ODdjZDc1NWQ4NTEwM2EyMzI3ZjA0NjE5ZWFlMTJhYzVhYzk4ZWVkZWVjOTM0ZGJmZmQ5ZDBhMDlkZTMzYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.TRHrkQ7akSKijZ7yK6XFSdhqxBIMn603YQ_Aguv4B-g)

Это не "в очередь на БГ встало 16 ботов из 10 возможных" а "в очередь на БГ встал 1 игрок и за ним 15 ботов, для этого БГ требуется всего минимум 10 человек"

> 6 лишних отсеялись !

> Почему эти 10 делятся на 2 команды по 5 ботов в каждую?

Успели зайти лишь по 5 ботов с каждой стороны, остальные отменились.
Боты для БГ абортируются только в 2 случаях:
1) Все игроки вышли из БГ и никого нет в инвайтах
2) БГ переполнено (для варсонга по умолчанию макс. 20 игроков)

Я вижу только 2 возможных варианта: либо максимум игроков изменяется скриптом, либо из-за ГМ режима (такой игрок не считается участником БГ)

**Comment by Bogir** [comment]
Пробую исключительно игровым персонажем.
ГМ во втором окне "на всякий пожарный".

Заметил недокументированную ситуацию:
после создания инста для выбранного БГ с ботами, как только персонаж оказался внутри стартовой комнаты, в течение 2 минут в этот же БГ-инст могут добавляться другие игроки.
Вот это просто супер! 
В этом случае половина свободных мест в командах вполне оправданы.

**Comment by trickerer** [comment]
Поставь в конфиге `Logger.bg.battleground=5,Console Server` и `Logger.npcbots=5,NpcBots Server`, и скинь полный лог из игры и из консоли сервера от попытки зайти в очередь на БГ до выхода (выйти можно после окончания таймера подготовки)

> Заметил недокументированную ситуацию:
после создания инста для выбранного БГ с ботами, как только персонаж оказался внутри стартовой комнаты, в течение 2 минут в этот же БГ-инст могут добавляться другие игроки.
Вот это просто супер!
В этом случае половина свободных мест в командах вполне оправданы.

Вообще это (постепенное добавление ботов) сделано для того, чтобы желающие игроки могли заполнить БГ до и вместо ботов. Если участников станет максимум остальные боты просто отменятся и будет БГ с максимально возможным количеством игроков

**Comment by Bogir** [comment]
лог из игры
![изображение](https://github.com/user-attachments/assets/8278632a-f417-4762-9e5c-58e70cd3cb67)

лог из консоли
```
24-10-2024 21:31:09 Online 2 players, 2 sessions
24-10-2024 21:32:10 Online 2 players, 2 sessions
24-10-2024 21:32:11 Adding Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 4, leader Etosham
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 2
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Adding BOT-Group to BattlegroundQueue bgTypeId: 2, bracket_id: 7, index: 3
24-10-2024 21:32:11 Generate bots for BG 2 for leader MapID: 571 X: 5739.42 Y: 671.678 Z: 644.862 O: 2.21763
GUID Full: 0x00000000000005ff Type: Player Low: 1535 Entry 0
Name: Etosham
AliveState: true UnitMovementFlags: 0 ExtraUnitMovementFlags: 0 Class: 7 (0 members)
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting ALL bots by Character id 1535 bg queue id 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71860 bg 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71868 bg 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71861 bg 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71869 bg 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71862 bg 2!
24-10-2024 21:32:34 BotBattlegroundEnterEvent: Aborting bot 71870 bg 2!
24-10-2024 21:33:11 Online 2 players, 2 sessions
24-10-2024 21:34:12 Online 2 players, 2 sessions
24-10-2024 21:35:13 Online 2 players, 2 sessions
```

**Comment by trickerer** [comment]
Думаю, проблема в конфиге, `Battleground.InvitationType` должно быть `0` Иначе менеджер очереди БГ отсеивает "лишних" ботов

**Comment by Bogir** [comment]
Оно так и установлено

```
#        Default:     0 - (Normal, Invite as much players to battlegrounds as queued,
#                          Don't bother with balance)
#                     1 - (Experimental, Don't allow to invite much more players
#                          of one faction)
#                     2 - (Experimental, Try to have even teams)

Battleground.InvitationType = 0
```

**Comment by trickerer** [comment]
Есть ли изменения в `battleground_template`? Чему равно `NpcBot.WanderingBots.BG.TargetTeamPlayersCount.WS` в конфиге?

Судя по логу в очередь встала одна группа + 16 ботов. Такое может быть только если в момент вычисления нужного колчества ботов группа, вставшая в очередь изначально имеет чётное количество участников, либо уже не в очереди (если `TargetTeamPlayersCount` = 8 то этот случай)

Ещё идея - поробовать поменять режим сервера с FFAPvP на PvP

**Comment by Bogir** [comment]
```
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.AV = 30
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.WS = 8
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.AB = 12
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.EY = 0
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.SA = 0
NpcBot.WanderingBots.BG.TargetTeamPlayersCount.IC = 0
```
